framework/src/base/Assembly.C

```diff
@@ -990,38 +990,57 @@ Assembly::init(const CouplingMatrix * cm)
 
   unsigned int n_vars = _sys.nVariables();
 
-  // I want the blocks to go by columns first to reduce copying of shape function in assembling
-  // "full" Jacobian
-  _cm_entry.clear();
+  _cm_ss_entry.clear();
+  _cm_sf_entry.clear();
+  _cm_fs_entry.clear();
+  _cm_ff_entry.clear();
 
-  const std::vector<MooseVariableFEBase *> & vars = _sys.getVariables(_tid);
+  auto & vars = _sys.getVariables(_tid);
 
-  for (const auto & jvar : vars)
+  _block_diagonal_matrix = true;
+  for (auto & ivar : vars)
   {
-    unsigned int j = jvar->number();
-    for (const auto & ivar : vars)
+    auto i = ivar->number();
+    ConstCouplingRow ccr(i, *_cm);
+    for (const auto & j : ccr)
     {
-      unsigned int i = ivar->number();
-      if ((*_cm)(i, j) != 0)
-        _cm_entry.push_back(std::make_pair(ivar, jvar));
+      if (_sys.isScalarVariable(j))
+      {
+        auto & jvar = _sys.getScalarVariable(_tid, j);
+        _cm_fs_entry.push_back(std::make_pair(ivar, &jvar));
+        _block_diagonal_matrix = false;
+      }
+      else
+      {
+        auto & jvar = _sys.getVariable(_tid, j);
+        _cm_ff_entry.push_back(std::make_pair(ivar, &jvar));
+        if (i != j)
+          _block_diagonal_matrix = false;
+      }
     }
   }
 
-  unsigned int max_rows_per_column = 0; // If this is 1 then we are using a block-diagonal
-                                        // preconditioner and we can save a bunch of memory.
-  for (unsigned int i = 0; i < n_vars; i++)
+  auto & scalar_vars = _sys.getScalarVariables(_tid);
+
+  for (auto & ivar : scalar_vars)
   {
-    unsigned int max_rows_per_this_column = 0;
-    for (unsigned int j = 0; j < n_vars; j++)
-    {
-      if ((*_cm)(i, j) != 0)
-        max_rows_per_this_column++;
-    }
-    max_rows_per_column = std::max(max_rows_per_column, max_rows_per_this_column);
+    auto i = ivar->number();
+    ConstCouplingRow ccr(i, *_cm);
+    for (const auto & j : ccr)
```

Comment:
You may want to construct the `ConstCouplingRow` object on the fly, since the `ccr` object is not used for anything else. For example:
```
for (const auto & j : ConstCouplingRow(i, *_cm))
```

Reply:
Will do